bucketupload/activity.go

```diff
@@ -0,0 +1,56 @@
+package bucketupload
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"go.artefactual.dev/tools/temporal"
+	"gocloud.dev/blob"
+)
+
+const Name = "bucket-upload"
+
+type (
+	Params struct {
+		Path       string
+		Key        string
+		BufferSize int
+	}
+	Result struct {
+		Key string
+	}
+	Activity struct {
+		bucket *blob.Bucket
+	}
+)
+
+func New(bucket *blob.Bucket) *Activity {
+	return &Activity{bucket: bucket}
+}
+
+func (a *Activity) Execute(ctx context.Context, params *Params) (*Result, error) {
+	h := temporal.StartAutoHeartbeat(ctx)
+	defer h.Stop()
+
+	file, err := os.Open(params.Path)
+	if err != nil {
+		return nil, fmt.Errorf("bucketupload: open file: %w", err)
+	}
+	defer file.Close()
+
+	if params.Key == "" {
+		params.Key = filepath.Base(params.Path)
```

Comment:
Mutating `params` might cause an unexpected results in the caller because it's a pointer - though to be honest I don't know how passing a pointer parameter to an activity works in Temporal considering the workflow is sending the payload to the worker via gRPC, as there is no shared memory location to point at.

Reply:
Good point, I doubt it will be an issue, but I'll use another var, just in case.